### PR TITLE
Feature/enterprise GitHub authentication

### PR DESCRIPTION
**Pull request by AlexisNeveu**
Enable Github and enterprise Github authentication on Parse Server.

### CODE REVIEW

File: src/auth_modules/enterpriseGithub.js [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+/**
+ * A promisey wrapper for api requests.
+ * @param { string } path - Path to make the request.
+ * @param { string } accessToken - Access token to be validated.
+ * @returns { Promise } - Resolved if access token is valid and rejected if not.
+ */
+function request(path, accessToken) {
+  return httpsRequest.get({
+    host: process.env.OAUTH_ENTERPRISE_API_HOST,
+    path: process.env.OAUTH_ENTERPRISE_API_PATH + path,
```
Comment (aboubakar14) [inline src/auth_modules/enterpriseGithub.js:15]: does it mean that for OAUTH_ENTERPRISE_API_PATH its mandatory to ends with "/" ?

File: src/services/ConfigurationService.js [hunk 2]
```diff
@@ -23,6 +23,11 @@ export function loadConfiguration() {
         classNames: ['Posts', 'Comments'], // List of classes to support for query subscriptions
       },
       entities: [User, Group],
+      auth: {
+        github: {
```
Comment (Zorin95670) [inline src/services/ConfigurationService.js:28]: put this in configuration file not in code.

File: src/auth_modules/enterpriseGithub.js (hunk 1 above)
Comment (AlexisNeveu) [reply src/auth_modules/enterpriseGithub.js]: done

File: dev.env [hunk 3]
```diff
@@ -2,3 +2,6 @@ DATABASE_URI="postgres://leto_admin:password@db:5432/leto_db"
 APP_ID=leto-modelizer-api-dev
```
[the remark this answers is not in the document]
Comment (AlexisNeveu) [reply dev.env:2]: done

File: src/auth_modules/enterpriseGithub.js [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+    headers: {
+      Authorization: `bearer ${accessToken}`,
+      'User-Agent': 'parse-server',
+    },
+  });
+}
+
+/**
+ * Returns a promise that fulfills iff this user id is valid.
```
[the remark this answers is not in the document]
Comment (AlexisNeveu) [reply src/auth_modules/enterpriseGithub.js]: done

File: src/cloud/cloud-functions.js [hunk 5]
```diff
@@ -23,3 +23,71 @@ export async function isGroupUnique(isExistingObject, objectName, sessionToken,
...
+ * Github and store his data in the Parse Server.
+ * @param {string} temporaryCode - Temporary code obtained when authenticated via
+ * GitHub/enterpriseGitHub endpoint and used to get the access_token.
+ * @param {object} Parse - The Parse object to use SDK functionalities.
+ * @param {object} axios - The axios lib to make requests.
+ * @param {*} Octokit - Octokit class to interact with the GitHub API
+ * @param {object} enterpriseServer36Admin - Used to configure Octokit with enterprise GitHub API.
+ * @returns {Parse.User} The authenticated User if valid otherwise an error.
+ */
+export async function thirdPartyAuth(
```
[the remark this answers is not in the document]
Comment (AlexisNeveu) [reply src/cloud/cloud-functions.js:80]: done

File: src/cloud/cloud-functions.js [hunk 6]
```diff
@@ -23,3 +23,80 @@ export async function isGroupUnique(isExistingObject, objectName, sessionToken,
...
+ * @param {object} enterpriseServer36Admin - Used to configure Octokit with enterprise GitHub API.
+ * @returns {Parse.User} The authenticated User in Parse format.
+ */
+async function apiRequest(accessToken, Parse, Octokit, enterpriseServer36Admin) {
+  const myAuthData = {
+    access_token: accessToken,
+  };
+  let octokit;
+  let provider;
+  if (process.env.GITHUB_OAUTH_TYPE !== 'enterprise') {
```
Comment (Zorin95670) [inline src/cloud/cloud-functions.js:41]: put GITHUB_OAUTH_TYPE in configuration

File: src/cloud/cloud-functions.js [hunk 7]
```diff
@@ -23,3 +23,80 @@ export async function isGroupUnique(isExistingObject, objectName, sessionToken,
...
+ * @param {*} Octokit - Octokit class to interact with the GitHub API
+ * @param {object} enterpriseServer36Admin - Used to configure Octokit with enterprise GitHub API.
+ * @returns {Parse.User} The authenticated User in Parse format.
+ */
+async function apiRequest(accessToken, Parse, Octokit, enterpriseServer36Admin) {
+  const myAuthData = {
+    access_token: accessToken,
+  };
+  let octokit;
+  let provider;
```
Comment (Zorin95670) [inline src/cloud/cloud-functions.js:40]: break line after

File: src/cloud/cloud-functions.js [hunk 8]
```diff
@@ -23,3 +23,80 @@ export async function isGroupUnique(isExistingObject, objectName, sessionToken,
...
+    });
+  } else {
+    provider = 'enterpriseGithub';
+    const OctokitEnterprise36 = Octokit.plugin(enterpriseServer36Admin);
+    octokit = new OctokitEnterprise36({
+      auth: myAuthData.access_token,
+      baseUrl: process.env.OAUTH_APP_API_BASE_URL,
+    });
+  }
+  return octokit.request('GET /user', {
```
Comment (Zorin95670) [inline src/cloud/cloud-functions.js:54]: break line before

File: src/cloud/cloud-functions.js [hunk 9]
```diff
@@ -23,3 +23,80 @@ export async function isGroupUnique(isExistingObject, objectName, sessionToken,
...
+    headers: {
+      'X-GitHub-Api-Version': '2022-11-28',
+    },
+  }).then(
+    async (resp) => {
```
Comment (Zorin95670) [inline src/cloud/cloud-functions.js:59]: rename to response

File: src/cloud/cloud-functions.js [hunk 10]
```diff
@@ -23,3 +23,80 @@ export async function isGroupUnique(isExistingObject, objectName, sessionToken,
...
+      myAuthData.id = resp.data.id;
+      const user = new Parse.User();
+      user.set('username', resp.data.login);
+      user.set('firstname', resp.data.name);
+      return user.linkWith(provider, { authData: myAuthData }, { useMasterKey: true });
```
Comment (Zorin95670) [inline src/cloud/cloud-functions.js:64]:
```js
return user.linkWith(
  provider, 
  { 
    authData: {
       id: response.data.id,
       access_token: accessToken,
    },
  }, 
  { useMasterKey: true },
);
```

File: src/cloud/cloud-functions.js [hunk 11]
```diff
@@ -23,3 +23,80 @@ export async function isGroupUnique(isExistingObject, objectName, sessionToken,
...
+    });
+  }
+  return octokit.request('GET /user', {
+    headers: {
+      'X-GitHub-Api-Version': '2022-11-28',
+    },
+  }).then(
+    async (resp) => {
+      myAuthData.id = resp.data.id;
+      const user = new Parse.User();
```
Comment (Zorin95670) [inline src/cloud/cloud-functions.js:61]: break line after

File: src/cloud/cloud-functions.js (hunk 7 above)
Comment (Zorin95670) [inline src/cloud/cloud-functions.js:40]: useless variable, like I sax before create a boolean in configuration, and when use it in ternary like:
configuration.isGithubPublic ? 'public' : 'enterpriseGithub' in the  `user.linkWith`

File: src/cloud/cloud-functions.js [hunk 12]
```diff
@@ -23,3 +23,80 @@ export async function isGroupUnique(isExistingObject, objectName, sessionToken,
...
+ */
+export async function thirdPartyAuth(
+  temporaryCode,
+  Parse,
+  axios,
+  Octokit,
+  enterpriseServer36Admin,
+) {
+  return axios.post(
+    process.env.OAUTH_APP_ACCESS_TOKEN_URL,
```
Comment (Zorin95670) [inline src/cloud/cloud-functions.js:88]: for all put in configuration

File: src/index.js [hunk 13]
```diff
@@ -18,6 +18,11 @@ app.get('/', (req, res) => {
 httpServer.listen(configuration.port, () => {
   console.log(`leto-modelizer-api running on port ${configuration.port}.`);
   console.log(`leto-modelizer-api running in mode ${configuration.mode}.`);
+  if (process.env.GITHUB_OAUTH_TYPE === 'enterprise') {
+    console.log('Enterprise GitHub authentication is enabled');
```
Comment (Zorin95670) [inline src/index.js:22]: Prefer: `Github authentication type: Public/Enterprise`

File: README.md [hunk 14]
```diff
@@ -91,6 +102,13 @@ DATABASE_URI=postgres://leto_admin:password@localhost:5432/leto_db
 APP_ID=my_app_id
 MASTER_KEY=my_super_secret_master_key
 MASTER_KEY_IPS=1.1.1.1/1,2.2.2.2/2,3.3.3.3/3
+OAUTH_APP_CLIENT_ID=0a000a000a0a0000a
+OAUTH_APP_CLIENT_SECRET=0a000a000a00aa0aaa0a0000a
+GITHUB_OAUTH_TYPE=enterprise
+OAUTH_APP_ACCESS_TOKEN_URL=https://myenterprisehost/login/oauth/access_token
+OAUTH_APP_API_BASE_URL=https://myenterprisehost/customghapi
+OAUTH_ENTERPRISE_HOST=myenterprisehost
```
Comment (Zorin95670) [inline README.md:110]: merge OAUTH_ENTERPRISE_HOST and OAUTH_ENTERPRISE_PATH

File: src/auth_modules/enterpriseGithub.js [hunk 15]
```diff
@@ -0,0 +1,49 @@
...
+
+/**
+ * Returns a promise that fulfills if this app id is valid.
+ * @returns { Promise } Resolved if this app id is valid and rejected if not.
+ */
+function validateAppId() {
+  return Promise.resolve();
+}
+
+export default {
```
Comment (Zorin95670) [inline src/auth_modules/enterpriseGithub.js:46]: change by class and add Configuration in constructor

File: src/services/ConfigurationService.js [hunk 16]
```diff
@@ -23,6 +24,14 @@ export function loadConfiguration() {
...
+          enabled: true,
+        },
+        enterpriseGithub: {
```
[the remark this answers is not in the document]
Comment (AlexisNeveu) [reply src/services/ConfigurationService.js:40]: I did not create a new class, I just pass the environement variable needed by enterpriseGithub in the constructor.

File: src/auth_modules/enterpriseGithub.js (hunk 15 above)
Comment (AlexisNeveu) [reply src/auth_modules/enterpriseGithub.js]: Done without passing the whole Configuration in constructor.

File: README.md (hunk 14 above)
Comment (AlexisNeveu) [reply README.md]: Done, I use only the OAUTH_APP_API_BASE_URL and split it when needed.

File: src/models/Configuration.js [hunk 17]
```diff
@@ -36,6 +38,7 @@ class Configuration {
       serverURL: 'http://localhost:1337/api',
       cloud: 'src/cloud/main.js',
```
[the remark this answers is not in the document]
Comment (AlexisNeveu) [reply src/models/Configuration.js]: done

File: src/index.js (hunk 13 above)
Comment (AlexisNeveu) [reply src/index.js]: done

File: src/cloud/cloud-functions.js (hunk 12 above)
Comment (AlexisNeveu) [reply src/cloud/cloud-functions.js]: done

File: src/cloud/cloud-functions.js (hunk 7 above)
Comment (AlexisNeveu) [reply src/cloud/cloud-functions.js]: done

File: src/cloud/cloud-functions.js (hunk 11 above)
Comment (AlexisNeveu) [reply src/cloud/cloud-functions.js:61]: done

File: src/cloud/cloud-functions.js (hunk 10 above)
Comment (AlexisNeveu) [reply src/cloud/cloud-functions.js]: done

File: src/cloud/cloud-functions.js [hunk 18]
```diff
@@ -23,3 +23,80 @@ export async function isGroupUnique(isExistingObject, objectName, sessionToken,
...
+/**
+ * This function makes the call to the GitHub/enterprise GitHub API to retrieve his information.
+ * @param {string} accessToken - Access token of the authenticated user.
+ * @param {object} Parse - The Parse object to use SDK functionalities.
+ * @param {*} Octokit - Octokit class to interact with the GitHub API
+ * @param {object} enterpriseServer36Admin - Used to configure Octokit with enterprise GitHub API.
+ * @returns {Parse.User} The authenticated User in Parse format.
+ */
+async function apiRequest(accessToken, Parse, Octokit, enterpriseServer36Admin) {
+  const myAuthData = {
```
[the remark this answers is not in the document]
Comment (AlexisNeveu) [reply src/cloud/cloud-functions.js]: done

File: src/cloud/cloud-functions.js (hunk 9 above)
Comment (AlexisNeveu) [reply src/cloud/cloud-functions.js]: done

File: src/cloud/cloud-functions.js (hunk 8 above)
Comment (AlexisNeveu) [reply src/cloud/cloud-functions.js:55]: done

File: src/cloud/cloud-functions.js [hunk 19]
```diff
@@ -23,3 +23,80 @@ export async function isGroupUnique(isExistingObject, objectName, sessionToken,
...
+  let octokit;
+  let provider;
+  if (process.env.GITHUB_OAUTH_TYPE !== 'enterprise') {
+    provider = 'github';
+    octokit = new Octokit({
+      auth: myAuthData.access_token,
+    });
+  } else {
+    provider = 'enterpriseGithub';
+    const OctokitEnterprise36 = Octokit.plugin(enterpriseServer36Admin);
```
[the remark this answers is not in the document]
Comment (AlexisNeveu) [reply src/cloud/cloud-functions.js:47]: done

File: src/cloud/cloud-functions.js (hunk 7 above)
Comment (AlexisNeveu) [reply src/cloud/cloud-functions.js]: done

File: src/cloud/cloud-functions.js (hunk 6 above)
Comment (AlexisNeveu) [reply src/cloud/cloud-functions.js]: done

Comment (AlexisNeveu) [reply src/cloud/cloud-functions.js]: Replaced by a boolean 'isEnterpriseGithub'

File: src/auth_modules/enterpriseGithub.js [hunk 20]
```diff
@@ -0,0 +1,62 @@
...
+  });
+}
+
+class EnterpriseGithub {
+  constructor(baseURL) {
+    if (baseURL) {
+      const splitURL = baseURL.toString().split('/');
+      // eslint-disable-next-line prefer-destructuring
+      this.host = splitURL[2];
+      for (let i = 3; i < splitURL.lenght; i += 1) {
```
Comment (Zorin95670) [inline src/auth_modules/enterpriseGithub.js:32]: Replace this by a regex

Comment (Zorin95670) [reply src/auth_modules/enterpriseGithub.js:32]: like :
```js
const result = /Regex with named groups/.exec(baseUrl)
this.host = result.groups.host;
this.endpoint = result.groups.endpoint;
```

File: src/utils/httpsRequest.js [hunk 21]
```diff
@@ -0,0 +1,43 @@
+/* eslint-disable */
```
Comment (Zorin95670) [inline src/utils/httpsRequest.js:1]: remove this and fix eslint error

File: README.md [hunk 22]
```diff
@@ -70,16 +74,21 @@ Now your project is up and running using Docker, providing a convenient and isol
...
+| MASTER_KEY_IPS             | No, default: [], no ip allowed             | 0.0.0.0/0                                             | Restrict masterKey to be used by only these ips. 0.0.0.0/0 enables any ip, not recommended for production.                                          |
+| OAUTH_APP_CLIENT_ID        | Yes                                        | 0a000a000a0a0000a                                     | The ID of the OAuth App you created on your own GitHub/Enterprise GitHub dashboard.                                                                 |
+| OAUTH_APP_CLIENT_SECRET    | Yes                                        | 0a000a000a00aa0aaa0a0000a                             | A secret token generated on the OAuth App you created on your own GitHub/Enterprise GitHub dashboard.                                               |
+| GITHUB_OAUTH_TYPE          | No, if not 'enterprise', public GH is used | enterprise                                            | Defines whether the app will use an enterprise GitHub authentication or the public one, if set to 'enterprise', the enterprise API URL is required. |
```
Comment (Zorin95670) [inline README.md:85]: No, if not 'enterprise', public GH is used -> No, only enterprise/public are accepted, default is public

File: src/services/ConfigurationService.js [hunk 23]
```diff
@@ -9,10 +9,19 @@ import Group from '../entities/Group.js';
  */
 export function loadConfiguration() {
   dotenv.config({ path: `${process.env.NODE_ENV}.env` });
+  let isEnterpriseGitHub = false;
```
Comment (aboubakar14) [inline src/services/ConfigurationService.js:13]: let isEnterpriseGitHub = process.env.GITHUB_OAUTH_TYPE === 'enterprise';

File: src/auth_modules/enterpriseGithub.js (hunk 20 above)
Comment (AlexisNeveu) [reply src/auth_modules/enterpriseGithub.js]: done

File: src/utils/httpsRequest.js (hunk 21 above)
Comment (AlexisNeveu) [reply src/utils/httpsRequest.js]: done

File: README.md (hunk 22 above)
Comment (AlexisNeveu) [reply README.md]: done

File: src/services/ConfigurationService.js (hunk 23 above)
Comment (AlexisNeveu) [reply src/services/ConfigurationService.js]: done

File: dev.env [hunk 24]
```diff
@@ -2,3 +2,10 @@ DATABASE_URI="postgres://leto_admin:password@db:5432/leto_db"
...
 MASTER_KEY=password
 MASTER_KEY_IPS=0.0.0.0/0
+OAUTH_APP_CLIENT_ID=MY_CLIENT_ID
+OAUTH_APP_CLIENT_SECRET=MY_CLIENT_SECRET
+# GITHUB_OAUTH_TYPE=public
+OAUTH_APP_ACCESS_TOKEN_URL=https://github.com/login/oauth/access_token
+# OAUTH_APP_API_BASE_URL=BASE_URL
+# OAUTH_ENTERPRISE_HOST=HOST
```
[the remark this answers is not in the document]
Comment (AlexisNeveu) [reply dev.env]: done

File: src/auth_modules/enterpriseGithub.js [hunk 25]
```diff
@@ -0,0 +1,61 @@
...
+      Authorization: `bearer ${accessToken}`,
+      'User-Agent': 'parse-server',
+    },
+  });
+}
+
+class EnterpriseGithub {
+  constructor(baseURL) {
+    if (baseURL) {
+      const { host, endpoint } = /^https:\/\/(?<host>.+)\/(?<endpoint>.+)$/.exec(
```
Comment (Zorin95670) [inline src/auth_modules/enterpriseGithub.js:29]: `/^https:\/\/(?<host>.+)\/(?<endpoint>.+)$/`
I think you can replace `(?<endpoint>.+)` by `(?<endpoint>[^/]+)`

File: src/utils/httpsRequest.js [hunk 26]
```diff
@@ -0,0 +1,61 @@
...
+
+/**
+ * Callback function for get and request methods
+ * @param {Function} resolve - Resolves the Promise where this function is called.
+ * @param {Function} reject - Rejects the Promise where this function is called.
+ * @param {boolean} noJSON - Wether the options in the request should be JSON or not.
+ * @returns {Promise} Promise with nothing on success otherwise an error.
+ */
+function makeCallback(resolve, reject, noJSON) {
+  return (res) => {
```
Comment (Zorin95670) [inline src/utils/httpsRequest.js:14]: res -> response

Comment (Zorin95670) [reply src/utils/httpsRequest.js:14]: or request not sure

File: src/utils/httpsRequest.js [hunk 27]
```diff
@@ -0,0 +1,61 @@
...
+import https from 'https';
+
+/**
+ * Callback function for get and request methods
+ * @param {Function} resolve - Resolves the Promise where this function is called.
+ * @param {Function} reject - Rejects the Promise where this function is called.
+ * @param {boolean} noJSON - Wether the options in the request should be JSON or not.
+ * @returns {Promise} Promise with nothing on success otherwise an error.
+ */
+function makeCallback(resolve, reject, noJSON) {
```
Comment (Zorin95670) [inline src/utils/httpsRequest.js:13]: add space inside the code of makeCallback to me more readable

File: src/utils/httpsRequest.js [hunk 28]
```diff
@@ -0,0 +1,61 @@
...
+
+/**
+ * A function to perfom a https request with callback.
+ * @param {*} options - JSON with request options or only the URL.
+ * @param {object} postData - data to pass when the method is POST.
+ * @returns {Promise} Promise with nothing on success otherwise an error.
+ */
+function request(options, postData) {
+  return new Promise((resolve, reject) => {
+    const req = https.request(options, makeCallback(resolve, reject));
```
Comment (Zorin95670) [inline src/utils/httpsRequest.js:54]: req -> request
and break line after

**Review by Zorin95670 — DISMISSED** [review]

**Review by Zorin95670 — APPROVED** [review]

**Review by aboubakar14 — APPROVED** [review]

**Pull request merged**